File: Java Blockchain/Multi/BlockchainMultiThreaded.java
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.Scanner; // Import the Scanner class
import java.util.concurrent.atomic.AtomicInteger;
import java.util.Date;
import java.util.List;



// blockchain class
public class BlockchainMultiThreaded{
	public List<Block> blockchain = new ArrayList<>();
    // Prefix dictates the difficulty of block mining, the higher the prefix, the more 0s required at the beginning of a block header hash
	static int prefix = 1;
	public volatile boolean nonceFound = false;
	public int nonce; 
	public int numThreads = 0;
	public boolean checked = false;
	public volatile boolean timeToCheck = false;
	public volatile boolean timeToCalc = false;
	public volatile int threadWhoFound = -1;


	// main method
	public static void main(String[] args) {
		BlockchainMultiThreaded chain = new BlockchainMultiThreaded();

		if (args.length > 0) {
			try {
				chain.numThreads = Integer.parseInt(args[0]);
				System.out.println("Num of threads " + chain.numThreads);
			} 
			catch (NumberFormatException e) 
			{
			}
		}
		else
		{
			chain.numThreads = 2;
		}

		// Print to screen creation of genesis block
		System.out.println("creating genesis block...");

		Scanner in = new Scanner(System.in);
		String data = null;
		// Grab the first file line as data for the genesis block
		data = in.nextLine();

		ArrayList<Double> times = new ArrayList<Double>();
		double shortestBlock = Double.MAX_VALUE;
		double longestBlock = Double.MIN_VALUE;
		double blockTime;
		double startBlock;
		double endBlock;

		// Instantiate the genesis block
		Block genesisBlock = new Block(data, "", new Date().getTime());
				
		// Create computers and put them in a list to keep track of them
        ArrayList<Computer> computers = new ArrayList<Computer>();
		ArrayList<Thread> threads = new ArrayList<Thread>();
		Thread thread; 
        for(int i = 0; i < chain.numThreads; i++)
        {
			Computer computer = new Computer(i+1, chain, prefix);
            computers.add(computer);
			computers.get(i).setBlock(genesisBlock);
			thread = new Thread(computer);
			threads.add(thread);
            thread.start();
        }
		startBlock = System.nanoTime();
		// Tell each thread to start calculating
		chain.timeToCalc = true;
		// Wait until we have possibly found the hash
		while(!chain.timeToCheck)
		{
		}

		genesisBlock.setNonce(computers.get(chain.threadWhoFound - 1).nonce);
		genesisBlock.setHash();
		
		// Validate our newly mined genesis block, using null as prev hash
		while(!genesisBlock.validateBlock(chain.blockchain, prefix))
		{
			System.out.println("lol rip 1");
			// Let the threads know the nonce has been checked and has not been found
			chain.timeToCheck = false;
			chain.nonceFound = false;

			// If our validation fails retry mining after resetting values
			genesisBlock.setNonce(0);
			genesisBlock.setTimeStamp(new Date().getTime());
			chain.nonce = 0;
			for(int i = 0; i < chain.numThreads; i++)
			{
				// Reset all necessary values
				computers.get(i).setBlock(genesisBlock);
				computers.get(i).prefix = prefix;
			}

			// Tell each thread to start calculating
			chain.timeToCalc = true;

			// Wait until we have possibly found the hash
			while(!chain.timeToCheck)
			{

			}

			genesisBlock.setNonce(computers.get(chain.threadWhoFound - 1).nonce);
			genesisBlock.setHash();

		}
		
		// Add genesis block to the blockchain
		chain.blockchain.add(genesisBlock);

		endBlock = System.nanoTime();
		//times.add(endBlock - startBlock);
		System.out.println("Blockchain initialized");
		
		System.out.println("Adding data blocks...");
		String previousHash = genesisBlock.getHash();
		
		
		// Continue until we reach the end of the input file
		while(in.hasNextLine()){

			// Read in line
			data = in.nextLine();
			startBlock = System.nanoTime();
			// Create a new block with our line of data
			Block newBlock = new Block(data, previousHash, new Date().getTime());
			chain.nonce = 0;
			// Give all threads our new block
			for(int i = 0; i < chain.numThreads; i++)
			{
				// Reset all necessary values
				computers.get(i).setBlock(newBlock);
				computers.get(i).prefix = prefix;
			}
			chain.nonceFound = false;
			chain.timeToCheck = false;

			// Tell each thread to start calculating
			chain.timeToCalc = true;

			// Wait until we have possibly found the hash
			while(!chain.timeToCheck)
			{

			}

			newBlock.setNonce(computers.get(chain.threadWhoFound - 1).nonce);
			newBlock.setHash();
			
			while(!newBlock.validateBlock(chain.blockchain, prefix))
			{

				System.out.println("lol rip" + newBlock.getNonce() + "  "+ newBlock.getHash() + " " + chain.threadWhoFound);
				// Let the threads know the nonce has been checked and has not been found
				chain.timeToCheck = false;
				chain.nonceFound = false;

				// If our validation fails retry mining after resetting values
				newBlock.setNonce(0);
				newBlock.setTimeStamp(new Date().getTime());
				chain.nonce = 0;
				for(int i = 0; i < chain.numThreads; i++)
				{
					// Reset all necessary values
					computers.get(i).setBlock(newBlock);
					computers.get(i).prefix = prefix;
				}

				// Tell each thread to start calculating
				chain.timeToCalc = true;

				// Wait until we have possibly found the hash
				while(!chain.timeToCheck)
				{
				}

				newBlock.setNonce(computers.get(chain.threadWhoFound - 1).nonce);
				newBlock.setHash();
			}

			// Add our newly mined and verified block and set up values for next block
			chain.blockchain.add(newBlock);
			endBlock = System.nanoTime();
			blockTime = endBlock - startBlock;
			times.add(blockTime);
			previousHash = newBlock.getHash();

		}
		for(int i = 0; i < chain.numThreads; i++)
		{
			computers.get(i).stop();
		}

		// Grab miminum time, maximum time, and total time
		double totalTime = 0.0; 
		for(Double time : times)
		{
			totalTime += time;
			if(shortestBlock > time)
			{
				shortestBlock = time;
			}
			if(longestBlock < time)
			{
				longestBlock = time;
			}
		}
		System.out.println("Blockchain Complete");
		System.out.println("Number of blocks added: " + chain.blockchain.size());
		System.out.println("Total execution time: " + String.format("%.4f",totalTime/ 1000000.00000)+ " ms.");
		String avg = String.format("%.4f", (double)(totalTime / (double)chain.blockchain.size())/ 1000000.00000 );
		System.out.println("Average block execution time: " + avg + " ms.");
		System.out.println("Fastest block execution time: " + String.format("%.4f", shortestBlock/ 1000000.00000) + " ms.");
		System.out.println("Slowest block execution time: " + String.format("%.4f",longestBlock/ 1000000.00000) + " ms.");
		in.close();

	}

    public synchronized int getNonce() {
        int temp = nonce;
        nonce = nonce + 1;
        return temp;
    }

	public synchronized void setBlockAndStop(int ID) {
		threadWhoFound = ID;
        timeToCheck = true;
		timeToCalc = false;
		nonceFound = true;
    }

}

// Block class
class Block{

	// Block properties
	public volatile String hash;
	private String previousHash;
	private String data;
	private long timeStamp;
	public AtomicInteger nonce;

	// Block constructor
	public Block(String data, String previousHash, long timeStamp){
		this.data = data;
		this.previousHash = previousHash;
		this.timeStamp = timeStamp;
    	this.nonce = new AtomicInteger(0);
		this.hash = calculateBlockHash();
	}

	// Standard getters
	public String getHash(){
		return this.hash;
	}

	public String getPreviousHash(){
		return this.previousHash;
	}

	public String getData(){
		return this.data;
	}

	public long timeStamp(){
		return this.timeStamp;
	}

	public AtomicInteger getNonce(){
		return this.nonce;
	}

	// Standard setters
	public synchronized void setHash(){
		this.hash = calculateBlockHash();
	}

	public void setPreviousHash(String previousHash){
		this.previousHash = previousHash;
	}

	public void setData(String data){
		this.data = data;
	}

	public void setTimeStamp(long timeStamp){
		this.timeStamp = timeStamp;
	}

	public synchronized void setNonce(int nonce){
		this.nonce.set(nonce);
	}

	// Method to calculate block hash
	public String calculateBlockHash(){
		// Concatenate parts of the block to generate the hash
		String dataToHash = previousHash
			+ Long.toString(timeStamp)
			+ Integer.toString(nonce.intValue())
			+ data;

        // Initialize the byte array
        byte[] byteRep = null;

        try {
            // Create a new MessageDigest object which implements SHA-256
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            // Generate the hash value of our input data, which is a byte array
            byteRep = md.digest(dataToHash.getBytes());
        } catch (Exception e) {
            e.printStackTrace();
        }
		// Convert byte array to correctly formatted string
        StringBuffer buffer = new StringBuffer();
    	for (byte b : byteRep) {
        	buffer.append(String.format("%02x", b));
    	}
    	return buffer.toString();
	}

	// Method to validate the block
	public boolean validateBlock(List<Block> blockchain, int prefix) {
		// Initialize a boolean value to true
		boolean valid = true;
	
		// The prefix string we want to match our hash prefix to
		String prefixString = new String(new char[prefix]).replace('\0', '0');
	
		// Local variable to store the size of the blockchain
		int size = blockchain.size();
	
		// If this blockchain is NOT empty then we are NOT validating the genesis block
		if(size >= 1){
		  // Validate the previousHash, hash and prefix
		  valid = (previousHash.equals(blockchain.get(size - 1).hash))
				&& (hash.equals(calculateBlockHash()))
				&& (hash.startsWith(prefixString));
			//System.out.println("prev hash  " + previousHash + " my prv hash" + blockchain.get(size - 1).hash);
			//System.out.println("my hash  " + hash + " recalced hash" + calculateBlockHash());
			//System.out.println("starting w prefix  " + hash.startsWith(prefixString));
		}
	
		// Otherwise it MUST be the genesis block since the blockchain is empty
		else {
		   valid = (previousHash.equals(""))
				&& (hash.equals(calculateBlockHash()))
				&& (hash.startsWith(prefixString));
	
		}
	
		// Return the value of valid
		return valid;
	}

}

class Computer implements Runnable{
    
	public boolean chainComplete = false;

	public Block currBlock;
    public int ID;
	private BlockchainMultiThreaded chain;
    public int nonce;
    

    public int prefix;

    public Computer(int ID, BlockchainMultiThreaded chain, int prefix) {
		this.ID = ID;
		this.chain = chain;
        this.prefix = prefix;
	}

    public void stop() {
		chainComplete = true;
	}

    public void run()
	{
		while(!chainComplete)
		{
			// Wait until it is time to calculate a hash
			if(chain.timeToCalc)
			{
                mineBlock(prefix);
                if(!chain.nonceFound)
                {
					chain.setBlockAndStop(ID);
                }
                else
                {
                    return;
                }
			}
            else
            {

            }
		}
	
    }

	

    // Method to mine a block
	public void mineBlock(int prefix){
		// Define the prefix we want to find
		String prefixString = new String(new char[prefix]).replace('\0', '0');
        nonce = chain.getNonce();
		// Until we find a hash beginning with the correct prefix, meaning we have found a hash smaller than our necessary target
		while(!calculateBlockHash().substring(0, prefix).equals(prefixString)){
			if(chain.nonceFound)
			{
				return;
			}
			// Grab the next nonce
            nonce = chain.getNonce();
			// Calculate another block hash
    	}

	}

    // Method to calculate block hash
	public String calculateBlockHash(){
		// Concatenate parts of the block to generate the hash
		String dataToHash = currBlock.getPreviousHash()
			+ Long.toString(currBlock.timeStamp())
			+ nonce
			+ currBlock.getData();

        // Initialize the byte array
        byte[] byteRep = null;

        try {
            // Create a new MessageDigest object which implements SHA-256
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            // Generate the hash value of our input data, which is a byte array
            byteRep = md.digest(dataToHash.getBytes());
        } catch (Exception e) {
            e.printStackTrace();
        }
		// Convert byte array to correctly formatted string
        StringBuffer buffer = new StringBuffer();
    	for (byte b : byteRep) {
        	buffer.append(String.format("%02x", b));
    	}
    	return buffer.toString();
	}

	public void setBlock(Block currentBlock)
	{
		this.currBlock = currentBlock;
	}
}

File: C++Blockchain/blockchainMultiThreaded.cpp
// Includes
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <thread>
#include <mutex>
#include <iterator>
#include <chrono>
#include <ctime>
#include <iomanip>
#include "sha256.h"

// Namespaces
using namespace std;

// Lock
mutex nonceLock;
int tempNonce = 0;

int getAndIncrementNonce()
{
    int currentNonce = tempNonce;
    tempNonce = currentNonce + 1;
    return currentNonce;
}

// Block class
class Block
{
public:
    // Block properties
    string data;
    string hash;
    string previousHash;
    long timeStamp;
    int nonce;
    bool foundHash;

    // Block constructor
    Block(string inputData, string inputPreviousHash, long inputTimeStamp)
    {
        this->data = inputData;
        this->previousHash = inputPreviousHash;
        this->timeStamp = inputTimeStamp;
        this->nonce = 0;
        this->hash = calculateBlockHash();
        this->foundHash = false;
    }

    // Standard getters
    string getHash()
    {
        return this->hash;
    }

    bool getFoundHash()
    {
        return this->foundHash;
    }

    string getPreviousHash()
    {
        return this->previousHash;
    }

    string getData()
    {
        return this->data;
    }

    long getTimeStamp()
    {
        return this->timeStamp;
    }

    int getNonce()
    {
        return this->nonce;
    }

    // Standard setters
    void setHash()
    {
        this->hash = calculateBlockHash();
    }

    void setFoundHash(bool input)
    {
        this->foundHash = input;
    }

    void setHashThreaded(string input)
    {
        this->hash = input;
    }

    void setPreviousHash(string inputHash)
    {
        this->previousHash = inputHash;
    }

    void setData(string inputData)
    {
        this->data = inputData;
    }

    void setTimeStamp(long inputTimeStamp)
    {
        this->timeStamp = inputTimeStamp;
    }

    void setNonce(int inputNonce)
    {
        this->nonce = inputNonce;
    }

    // Method to calculate block hash
    string calculateBlockHash()
    {
        // Concatenate parts of the block to generate the hash
        string dataToHash = this->previousHash + to_string(this->timeStamp) + to_string(this->nonce) + this->data;

        // Calculate hash based on block's data
        string hashedData = sha256(dataToHash);

        // Return hashed data
        return hashedData;
    }

    // Method to calculate block hash (Input parameters)
    string calculateBlockHashThreaded(string prevHash, long timeS, int non, string da)
    {
        // Concatenate parts of the block to generate the hash
        string dataToHash = prevHash + to_string(timeS) + to_string(non) + da;

        // Calculate hash based on block's data
        string hashedData = sha256(dataToHash);

        // Return hashed data
        return hashedData;
    }

    // Function that each thread runs
    void threadMineBlock(int threadID, int prefix)
    {
        // Prefix to look for in hash
        string prefixString(prefix, '0');

        // Search for hash with pre-determined prefix
        int i = 0;
        while(this->getFoundHash() == false)
        {
            // Get nonce
            nonceLock.lock();
            int localNonce = getAndIncrementNonce();
            nonceLock.unlock();

            // Calculate new hash
            string localHash = calculateBlockHashThreaded(this->previousHash, this->timeStamp, localNonce, this->data);

            // Check if hash contains the prefix
            if (localHash.substr(0, prefix) == prefixString && this->foundHash == false)
            {
                this->setFoundHash(true);
                this->setHashThreaded(localHash);
                this->setNonce(localNonce);
            }
        }
    }

    // Method to mine a block
    void mineBlock(int prefix)
    {
        // Get maximum number of threads the system can support
        int numThreads = thread::hardware_concurrency();

        // Create threads to mine block
        vector<thread> vecOfThreads;
        this->setFoundHash(false);
        tempNonce = 0;
        for (int i = 0; i < numThreads; i++)
        {
            thread threadObject(&Block::threadMineBlock, this, i, prefix);
            vecOfThreads.push_back(move(threadObject));
        }

        // wait for block to be mined
        while (foundHash == false)
        {
        }

        for (thread & th : vecOfThreads)
        {
            if (th.joinable())
            {
                th.join();
            }
        }

        // Finished mining the block
        return;
    }

    // Validate the block
    bool validateBlock(vector<Block> blockchain, int prefix)
    {
        // Initialize a boolean value to true
        bool valid = true;

        // The prefix string we want to match our hash prefix to
        string prefixString(prefix, '0');

        // Local variable to store the size of the blockchain
        int size = blockchain.size();

        // If this blockchain is NOT empty then we are NOT validating the genesis block
        if(size >= 1)
        {
            // Validate the previous hash, hash and prefix
            bool previousHashValid = this->getPreviousHash() == blockchain.at(size - 1).hash;
            bool hashValid = this->hash == calculateBlockHashThreaded(this->getPreviousHash(), this->getTimeStamp(), this->getNonce(), this->getData());
            // bool hashValid = this->hash == calculateBlockHash();
            valid = previousHashValid && hashValid;
        }
            // Otherwise it MUST be the genesis block since the blockchain is empty
        else
        {
            string emptyHash = "0000000000000000000000000000000000000000000000000000000000000000";
            bool previousHashValid = this->getPreviousHash() == emptyHash;
            bool hashValid = this->hash == calculateBlockHashThreaded(this->getPreviousHash(), this->getTimeStamp(), this->getNonce(), this->getData());
            // bool hashValid = this->hash == calculateBlockHash();
            valid = previousHashValid && hashValid;
        }

        // Return whether the block is valid or not.
        return valid;
    }
private:
};

// Main method
int main(int argc, char *argv[])
{
    // Get file path for input file
    string filePath = "../gorgias.txt";
    if (argv[1] != NULL)
    {
        filePath = argv[1];
    }

    // Get difficulty of block mining. Higher is harder (Default = 2)
    int prefix = 2;
    if (argv[2] != NULL)
    {
        prefix = atoi(argv[2]);
    }

    // Determine whether to output alldata or only test data (Default = test data only)
    bool fullOutput = false;
    if (argv[3] != NULL)
    {
        fullOutput = true;
    }

    // Create a vector of block times
    vector<float> executionTimes;

    // Start timer
    long startBlockchain = (chrono::duration_cast<chrono::microseconds>((chrono::time_point_cast<chrono::microseconds>(chrono::system_clock::now())).time_since_epoch())).count();

    // Create blockchain (list of blocks)
    vector<Block> blockchain;

    // Grab the first file line as data for the genesis block
    ifstream inputFile;
    inputFile.open(filePath, ios::in);
    string data;
    if (inputFile.is_open())
    {
        getline(inputFile, data);
    }
    else
    {
        if (fullOutput)
        {
            cout << "[Error]: FILE IS NOT OPEN 223" << endl;
        }
    }

    // Instantiate the genesis block
    long currentDate = (chrono::duration_cast<chrono::milliseconds>((chrono::time_point_cast<chrono::milliseconds>(chrono::system_clock::now())).time_since_epoch())).count();
    string emptyHash = "0000000000000000000000000000000000000000000000000000000000000000";
    Block genesisBlock(data, emptyHash, currentDate);

    if (fullOutput)
    {
        cout << "Creating genesis block" << endl;
    }

    // Mine the genesis block to create the hash for the next block.
    genesisBlock.mineBlock(prefix);

    if (fullOutput)
    {
        cout << "Validating genesis block" << endl;
    }

    // Validate our newly mined genesis block, using null as prev hash.
    while(!genesisBlock.validateBlock(blockchain, prefix))
    {
        // If our validation fails, retry mining after resetting values.
        genesisBlock.setNonce(0);
        long currentDate = (chrono::duration_cast<chrono::milliseconds>((chrono::time_point_cast<chrono::milliseconds>(chrono::system_clock::now())).time_since_epoch())).count();
        genesisBlock.setTimeStamp(currentDate);
        genesisBlock.calculateBlockHash();
        genesisBlock.mineBlock(prefix);
    }

    // Add genesis block to the blockchain
    blockchain.push_back(genesisBlock);

    if (fullOutput)
    {
        cout << "Blockchain initialized" << endl << "Adding data blocks..." << endl;
    }

    // Continue adding blocks until we reach the end of the input file
    while(inputFile.peek() != EOF)
    {
        // Start timer for this block
        long startBlock = (chrono::duration_cast<chrono::microseconds>((chrono::time_point_cast<chrono::microseconds>(chrono::system_clock::now())).time_since_epoch())).count();

        // Read in the next line
        if (inputFile.is_open())
        {
            getline(inputFile, data);
        }
        else
        {
            if (fullOutput)
            {
                cout << "[Error]: FILE IS NOT OPEN" << endl;
            }
        }

        // Create a new block with our line of data.
        long currentDate = (chrono::duration_cast<chrono::milliseconds>((chrono::time_point_cast<chrono::milliseconds>(chrono::system_clock::now())).time_since_epoch())).count();
        Block newBlock(data, blockchain.at(blockchain.size() - 1).getHash(), currentDate);

        // Mine the new block
        newBlock.mineBlock(prefix);

        // Validate new block
        while(!newBlock.validateBlock(blockchain, prefix))
        {
            // If our validation fails, retry mining after resetting values.
            newBlock.setNonce(0);
            long currentDate = (chrono::duration_cast<chrono::milliseconds>((chrono::time_point_cast<chrono::milliseconds>(chrono::system_clock::now())).time_since_epoch())).count();
            newBlock.setTimeStamp(currentDate);
            newBlock.calculateBlockHash();
            newBlock.mineBlock(prefix);
        }

        // Add our newly mined and verified block and set up values for next block
        blockchain.push_back(newBlock);

        // End timer for this block
        long endBlock = (chrono::duration_cast<chrono::microseconds>((chrono::time_point_cast<chrono::microseconds>(chrono::system_clock::now())).time_since_epoch())).count();

        // Calculate execution time for block
        float durationBlock = (endBlock - startBlock) / 1000.f;

        // Add duration to vector
        executionTimes.push_back(durationBlock);
    }

    // Close input file
    inputFile.close();

    // Stop timer
    long stopBlockchain = (chrono::duration_cast<chrono::microseconds>((chrono::time_point_cast<chrono::microseconds>(chrono::system_clock::now())).time_since_epoch())).count();
    if (fullOutput)
    {
        cout << "All blocks added!" << endl << endl;
    }

    // Calculate total execution time
    float durationBlockchain = (stopBlockchain - startBlockchain) / 1000.f;

    // Print block execution times to txt file & calculate interesting data features
    float sum = 0.0;
    float averageBlockTime = 0;
    float fastestBlockTime = 100000000.0;
    float slowestBlockTime = 0.0;

    for (int dataEntryIndex = 0; dataEntryIndex < executionTimes.size(); dataEntryIndex++)
    {

        // Add execution time to sum
        sum += executionTimes.at(dataEntryIndex);

        // Get fastest execution time
        if (executionTimes.at(dataEntryIndex) < fastestBlockTime)
        {
            fastestBlockTime = executionTimes.at(dataEntryIndex);
        }

        // Get slowest execution time
        if (executionTimes.at(dataEntryIndex) > slowestBlockTime)
        {
            slowestBlockTime = executionTimes.at(dataEntryIndex);
        }
    }

    // Calculate average
    averageBlockTime = sum / (executionTimes.size());

    // Display interesting data points.
    if (fullOutput)
    {
        cout << "=================== Data ===================" << endl;
        cout << setw(32) << left << "Number of blocks added:" << right << to_string(executionTimes.size()) << endl;
        cout << setw(32) << left << "Total execution time:" << right << to_string(durationBlockchain) + " ms." << endl;
        cout << setw(32) << left << "Average block execution time:" << right << to_string(averageBlockTime) + " ms." << endl;
        cout << setw(32) << left << "Fastest block execution time:" << right << to_string(fastestBlockTime) + " ms." << endl;
        cout << setw(32) << left << "Slowest block execution time:" << right << to_string(slowestBlockTime) + " ms." << endl;
    }
    else
    {
        cout << durationBlockchain << " " << averageBlockTime << endl;
    }

    return 0;
}
